[WIP] add min/max limits to temp plot (closes #240)

mlp6 opened the pull request:
@suyashkumar completely unsure if this is the correct way to this... almost definitely not the most elegant, or even remotely so, but hey... its JavaScript, what do you expect.  ;)  Please correct.

mlp6 committed:
add min/max limits to temp plot (closes #240)

mlp6 committed:
fix js syntax

mlp6 committed:
use tempMin/Max vars

suyashkumar commented on software/cloud-dashboard/react-client/src/components/temp-view/temp-plot.js:
```diff
@@ -38,6 +40,22 @@ class TempPlot extends Component {
 		return array;
 	};
 	
+  filterDataMinMax = (array) => {
```
This seems reasonable, but for the implementation of this function, consider using the `map` function as described [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) which will allow you to apply a function _f_ to each element _e_ in an array with the output _f(e)_ populating the corresponding index in an output array.

mlp6 replied:
Yeah, I wanted to use map... But I wasn't in Python. Syntax.  Bah.

suyashkumar commented on software/cloud-dashboard/react-client/src/components/temp-view/temp-plot.js:
```diff
@@ -9,6 +9,8 @@ import Input from 'react-toolbox/lib/input';
 const LINE_COLORS = ['#e41a1c', '#377eb8', '#4daf4a', '#984ea3', '#ff7f00', '#ffff33']
```
also, as seen [here](https://github.com/mlp6/ADPL/blob/master/software/cloud-dashboard/react-client/src/components/temp-view.js#L48) the `TempPlot` component has actually been replaced by a more generalized `Plot` component which is used to render both the temperature plot and the bucket view plot so these changes should likely go into the `plot` component! My apologies for not promptly removing this legacy component right away. Since it seems like the `Plot` component is working fine I can go ahead and remove this component once I ensure there are no lingering dependencies.

suyashkumar commented on software/cloud-dashboard/react-client/src/components/temp-view/temp-plot.js:
```diff
@@ -9,6 +9,8 @@ import Input from 'react-toolbox/lib/input';
 const LINE_COLORS = ['#e41a1c', '#377eb8', '#4daf4a', '#984ea3', '#ff7f00', '#ffff33']
 
 const defaultNumberOfPoints = 300;
+const tempMin = 0;
```
since these changes will be in the generalized `plot` component as described above, these should likely be input `props` to the component (perhaps `minLimit` and `maxLimit`). These values can be passed into the `Plot` component by the parent, as is seen [here](https://github.com/mlp6/ADPL/blob/master/software/cloud-dashboard/react-client/src/components/plot-card.js#L35). Inside the `Plot` component, you can of course access the current prop value using syntax similar to the following `this.props.minLimit`

mlp6 replied:
Would it be worth putting these in `constants.js`?  I'll take a stab at using the correct class for this now...

mlp6 replied:
oh wait, this plot component is also used for the bucket tips... that wouldn't make sense as a default for that.

suyashkumar reviewed (commented):
Overall LGTM, but had some comments about using `map` and moving over these changes to the new more generalized `Plot` component! I should have removed the temp-plot component earlier once it was successfully switched out with the more general Plot component, my apologies for the delay there.

mlp6 committed:
rm legacy temp-plot component (replaced by more generic plot component)

mlp6 committed:
add limits to YAxis in plot component

* Aaron requested limits on the temperature y-axis plots ranging from
  [0:80] degrees.  Implemented this through a YAxis domain option that
  will impose these hard limits if the data exceed these; otherwise the
  y-axis will be dynamically set based on the data.
* Chose this as opposed to re-mapping the data points that fall outside
  of this range so that the temperature data array integrity is
  maintained.

mlp6 commented:
Making some attempted changes just to get more familiar with the code base.  Latest commit imposes `YAxis` limits, but I haven't captured setting this for the bucket tips yet (still trying to find where those data are processed right now).  Obviously still a WIP...

mlp6 committed:
update test to use Plot component

mlp6 committed:
change TempPlot -> Plot component

mlp6 commented:
This is broken right now... so definitely do not merge this in!

mlp6 committed:
fix random things that probably do not matter

mlp6 committed:
fix ; syntax

mlp6 commented on software/cloud-dashboard/react-client/src/components/temp-view.js:
```diff
@@ -17,13 +17,20 @@ class TempView extends Component {
             tempData = tempData.reverse()
         }
+
+    this.props.dataMin = (tempData.min() < tempDataMin ? tempDataMin : tempData.min());
```
@suyashkumar trying to set these here seems to break the plots... insights?

mlp6 replied:
by breaking the plots, they just don't render

mlp6 replied:
nvm - I realize this is broken b/c of trying to assign something to `props` incorrectly (I think)

suyashkumar replied:
ah yes, you should not be setting this components own props from within it -- the parent to this component passes in props to this component somewhat like function inputs. 

Likely what you want to be doing is passing in `dataMin` and `dataMax` as input props to the plot itself right? In which case you would be passing them into `PlotCard` like `data` is being passed in on L56.

mlp6 replied:
yep... I realized this was a broken approach.  not surprised by that.

mlp6 committed:
rm legacy temp-plot component (replaced by more general plot)

mlp6 committed:
add yAxisMinMax array for plots

mlp6 committed:
merge in temp plot test branch with this one

mlp6 committed:
use Math.min/max() since my array methods do not actually seem to exist

mlp6 commented:
@suyashkumar help... in general things were working with this approach to both have the temp and bucket plots have appropriate y-axis scaling, but in refactoring the code, something happened where now I get plot timeouts.  before I spend more time figuring this out, can you vet the approach to setting the y-axis limits this way?

mlp6 commented:
also, the unit test is broken, but I think that it was calling things that the new `plot` component doesn't use anymore

suyashkumar commented on software/cloud-dashboard/react-client/src/components/plot.js:
```diff
@@ -92,7 +92,7 @@ class Plot extends Component {
                 <LineChart data={dataToShow}>
                     <XAxis dataKey="time" label="Date" interval={tickInterval} tickFormatter={this.formatDate} />
-                    <YAxis />
+                    <YAxis type="number" domain={[this.props.yAxisMinMax[0], this.props.yAxisMinMax[1]]}/>
```
`props.yAxisMinMax` is not available here yet, it needs to be passed into this `Plot` component from the `PlotCard` component and you should be good to use it!

mlp6 replied:
Fixed

mlp6 committed:
add yAxisMinMax to plot-card component

mlp6 commented:
Plots are now rendering, though the plots don't seem to be respecting the actual min/max limits correctly.  That sometimes seems to change as a function of when the plot is getting reloaded (?).  Not sure...

mlp6 commented:
Two suggestions:
1. We might want to have the plot min/max have the ability to be updated on the web page and default to values instead of just being hard-coded.
2. The averages that are displayed still use the raw "bad" data points, which probably needs to also be fixed with some data scrubbers.

Thoughts @aforbis-stokes @Graham-H-Miller

suyashkumar commented:
hmm may have to look deeper into that later, `console.log`ing the min/max values may help?

mlp6 commented:
figured it out, but will need some time to fix...

mlp6 committed:
fix calculation of yAxisMinMax values

mlp6 commented:
@suyashkumar I fixed the caluclation of the min/max y-axis values... they are properly console logged by from the `plot` component, but the actual rendered plot does not appear to be respecting the `<Yaxis ...` attribute that I am setting: https://github.com/mlp6/ADPL/blob/6f592390976a0dbe50335f1ffc63668bc21b1f9e/software/cloud-dashboard/react-client/src/components/plot.js#L95  

This could use your eyes when you have a chance...

mlp6 commented on software/cloud-dashboard/react-client/src/components/temp-view.js:
```diff
@@ -17,13 +17,26 @@ class TempView extends Component {
+    var tempDataMax = 80;
+
+    var allTempData = [];
+    for (var i in tempData) {
```
@suyashkumar Better way to do this in Javascript?

mlp6 commented on software/cloud-dashboard/react-client/src/containers/bucket-view.js:
```diff
@@ -72,6 +72,12 @@ class BucketView extends Component {
         }
 
+    var flowDataValues = [];
+    for (var i in flowData) {
```
@suyashkumar better way to do this in Javascript?

mlp6 commented on software/cloud-dashboard/react-client/src/containers/bucket-view.js:
```diff
@@ -72,6 +72,12 @@ class BucketView extends Component {
+    for (var i in flowData) {
+      flowDataValues[i] = flowData[i].flow;
+    }
+    var yAxisMinMax = [Math.min.apply(Math, flowDataValues), Math.max.apply(Math, flowDataValues)];
```
@suyashkumar do you really have to do `Math.min.apply(Math...` to work with an array?!

mlp6 committed:
fix y-axis plotting issue

It appears the recharts will override your YAxis domain unless you
explicitly set `allowDataOverflow=True`.

aforbis-stokes commented:
In regards to the hard limit/data scrubbing, voiding the faulty data would be better in my opinion for the reason you suggested for skewing averages. 
That said, a quick-fix of hard limits is ok with me for the moment if data scrubbing is turning out a challenge. The plot shared on Slack does look not-ideal, but we recently came across a rare connector problem which has now been fixed. Most data that would be discarded is only a few points across a 3-day span and would not cause so many lines entering/leaving the limits.

mlp6 commented:
@aforbis-stokes what would you like bad data points set to, just `NaN`?  I'm hesitating to set them to 0 or 80 since if they are bad, there is no reason to give them actual value.  I need to see what the `recharts` package that we're using for plotting will do with `NaN` values.

aforbis-stokes commented:
I do think NaN is better than setting to 0 or 80. There is some value in knowing that data is funky, and the extreme numbers we see are good indicators of hardware issues. Is it possible to get something like "-NaN" for <0 and "Nan" or "+NaN" for >80? Typically a negative value means a probe is disconnected, while a very large number means things are connected, but something is damaged.

mlp6 commented:
ok - I don't think that we can assign a sign to the `NaN`, but what we can do is when we scrub the data we can look for those trends and determine set a status for that thermistor based on the sign of the bad values and record those in the logs.

aforbis-stokes commented:
That sounds good to me.

mlp6 committed:
clean up ESLint warnings

mlp6 committed:
rm what appears to be an unused import

mlp6 commented:
@suyashkumar will also need help updating the unit tests that you have for `temp-view`.  Getting errors associated with:
1. https://github.com/mlp6/ADPL/blob/82dc6720ae2049c2e22b17bbadbd02dae2849741/software/cloud-dashboard/react-client/src/test/components/temp-view/temp-view.test.js#L16
2. https://github.com/mlp6/ADPL/blob/82dc6720ae2049c2e22b17bbadbd02dae2849741/software/cloud-dashboard/react-client/src/test/components/temp-view/temp-view.test.js#L20

I'm flying completely blind on what you are actually testing with those, but there appears to be something that has internally changed when you went from a `temp-plot` component to the more generalized `plot` component.

mlp6 committed:
replace clunky loops with array reducers

mlp6 committed:
add console logs to help debug tricky arrays
